### PR TITLE
[mesheryctl] Add support to update spreadsheet during model generation.

### PR DESCRIPTION
**Pull request by MUzairS15**
**Notes for Reviewers**

This PR fixes #

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.

### CODE REVIEW

File: mesheryctl/internal/cli/root/registry/error.go [hunk 1]
```diff
@@ -19,7 +19,7 @@ func ErrUpdateRegistry(err error, path string) error {
 }
 
 func ErrGenerateModel(err error, modelName string) error {
-	return errors.New(ErrGenerateModelCode, errors.Alert, []string{fmt.Sprintf("error generating model %s", modelName)}, []string{err.Error()}, []string{"Registrant used for the model is not available", "Failed to create model directory"}, []string{"Check network connevtivity and try again.", "Ensure sufficient permissions to allow creation of model directory"})
+	return errors.New(ErrGenerateModelCode, errors.Alert, []string{fmt.Sprintf("error generating model: %s", modelName)}, []string{err.Error()}, []string{"Registrant used for the model is not supported", "Verify the source url", "Failed to create model directory"}, []string{"Ensure one of the supported registrant is used for model generation", "Ensure correct source url is provided and in correct format", "Ensure sufficient permissions to allow creation of model directory"})
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/registry/error.go:22]: 1. registrants - plural

2. The registrant is not performing model generation itself. That sentence needs repressing.

File: mesheryctl/internal/cli/root/registry/generate.go [hunk 2]
```diff
@@ -219,7 +222,7 @@ func InvokeGenerationFromSheet() error {
 				utils.Log.Error(ErrGenerateModel(err, model.Model))
 				return
 			}
-			utils.Log.Info("Extracted ", len(comps), "for model: %s", model.ModelDisplayName)
+			utils.Log.Info("Extracted ", len(comps), "components for model: %s", model.ModelDisplayName)
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/registry/generate.go:225]: Missing whitespace before components.

File: mesheryctl/internal/cli/root/registry/generate.go [hunk 3]
```diff
@@ -309,6 +318,31 @@ func GenerateDefsForCoreRegistrant(model utils.ModelCSV) error {
...
+	if err != nil {
+		return nil, ErrGenerateModel(err, "unable to start model generation")
+	}
+	return modelCSVHelper, nil
+}
+
+func parseComponentSheet(url string) (*utils.ComponentCSVHelper, error) {
+	compCSVHelper, err := utils.NewComponentCSVHelper(url, "Components", componentSpredsheetGID)
+	if err != nil {
+		return nil, err
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/registry/generate.go:337]: Missing specific error message.

File: mesheryctl/pkg/utils/component.go [hunk 4]
```diff
@@ -94,7 +95,7 @@ type ComponentCSVHelper struct {
 
 func NewComponentCSVHelper(sheetURL, spreadsheetName string, spreadsheetID int64) (*ComponentCSVHelper, error) {
 	sheetURL = sheetURL + "/pub?output=csv" + "&gid=" + strconv.FormatInt(spreadsheetID, 10)
-	fmt.Println("Downloading CSV from:", sheetURL)
+	Log.Info("Downloading CSV from:", sheetURL)
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/component.go:98]: Missing whitespace after colon.

File: mesheryctl/pkg/utils/component.go [hunk 5]
```diff
@@ -152,9 +153,9 @@ func (mch *ComponentCSVHelper) ParseComponentsSheet() error {
 				mch.Components[data.Registrant][data.Model] = make([]ComponentCSV, 0)
 			}
 			mch.Components[data.Registrant][data.Model] = append(mch.Components[data.Registrant][data.Model], data)
-			fmt.Printf("Reading Component: %s for Modal: %s from Registrant: %s\n", data.Component, data.Model, data.Registrant)
+			Log.Info(fmt.Sprintf("Reading Component: %s for Model: %s from Registrant: %s\n", data.Component, data.Model, data.Registrant))
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/component.go:156]:
```suggestion
			Log.Info(fmt.Sprintf("Reading Registrant [%s] Model [%s] Component [%s]\n", data. Registrant, data.Model, data. Component))
```

Ordering the data this way will be easier to scan with your eyes.

File: mesheryctl/pkg/utils/component.go [hunk 6]
```diff
@@ -261,10 +262,22 @@ func CreateComponentsMetadataAndCreateSVGsForMDStyle(components []ComponentCSV,
 
 func (m ComponentCSVHelper) Cleanup() error {
 	// remove csv file
-	fmt.Println("Removing CSV file:", m.CSVPath)
+	Log.Info("Removing CSV file:", m.CSVPath)
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/component.go:265]:
```suggestion
	Log.Info("Removing CSV file: ", m.CSVPath)
```

File: mesheryctl/pkg/utils/component.go [hunk 7]
```diff
@@ -261,10 +262,22 @@ func CreateComponentsMetadataAndCreateSVGsForMDStyle(components []ComponentCSV,
...
 		return err
 	}
 	return nil
 }
+
+func ConvertCompDefToCompCSV(modelcsv *ModelCSV, compDef v1alpha1.ComponentDefinition) *ComponentCSV {
+	return &ComponentCSV{
+		Registrant:       modelcsv.Registrant,
+		Model:            modelcsv.Model,
+		Component:        compDef.Kind,
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/component.go:277]: Please don’t hardcode the value of annotation.

File: mesheryctl/pkg/utils/component.go [hunk 8]
```diff
@@ -261,10 +262,22 @@ func CreateComponentsMetadataAndCreateSVGsForMDStyle(components []ComponentCSV,
...
+		IsAnnotation:     "false",
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/component.go:278]: Please don’t hardcode the value of annotation.

File: mesheryctl/pkg/utils/error.go [hunk 9]
```diff
@@ -485,3 +487,19 @@ func ErrRequestResponse(err error) error {
...
+		[]string{"Error occured converting struct to [][]string", "The struct might have in-appropriate csv tags"},
+		[]string{"Verify the struct is non-empty", "Verify the struct has appropriate csv tags"})
+}
+
+func ErrAppendToSheet(err error, id string) error {
+	return errors.New(ErrAppendToSheetCode, errors.Alert,
+		[]string{fmt.Sprintf("Failed to append data into sheet %s", id)},
+		[]string{err.Error()},
+		[]string{"Error occured while appending to the spreadsheet", "The credentials might be incorrect/expired"},
+		[]string{"Ensure correct append range (A1 notation) is used", "Ensure correct credentials are used"})
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/error.go:504]: How many credentials are they using to access the spreadsheet? If just one, then rephrase.

File: mesheryctl/pkg/utils/error.go [hunk 10]
```diff
@@ -485,3 +487,19 @@ func ErrRequestResponse(err error) error {
 		[]string{"Error occurred while generating a response"},
 		[]string{"Check your network connection and the status of Meshery Server via `mesheryctl system status`."})
 }
+
+func ErrMarshalStructToCSV(err error) error {
+	return errors.New(ErrMarshalStructToCSVCode, errors.Alert,
+		[]string{"Failed to marshal struct to csv"},
+		[]string{err.Error()},
+		[]string{"Error occured converting struct to [][]string", "The struct might have in-appropriate csv tags"},
+		[]string{"Verify the struct is non-empty", "Verify the struct has appropriate csv tags"})
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/error.go:496]: This is meaningless to the user. Offer something that they can understand and do something about, like stating that it’s possible that the column names in the spreadsheet are misconfigured and don’t align with the required template format.

File: mesheryctl/internal/cli/root/registry/generate.go [hunk 11]
```diff
@@ -309,6 +318,31 @@ func GenerateDefsForCoreRegistrant(model utils.ModelCSV) error {
 	return nil
 }
 
+func parseModelSheet(url string) (*utils.ModelCSVHelper, error) {
+	modelCSVHelper, err := utils.NewModelCSVHelper(url, "Models", sheetGID)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (MUzairS15) [reply mesheryctl/internal/cli/root/registry/generate.go:324]: The specifics of the error is being returned from the function `NewModelCSVHelper`

File: mesheryctl/internal/cli/root/registry/generate.go (hunk 3 above)
Comment (MUzairS15) [reply mesheryctl/internal/cli/root/registry/generate.go:337]: The specifics of the error is being returned from the function NewComponentCSVHelper

File: mesheryctl/internal/cli/root/registry/error.go [hunk 12]
```diff
@@ -15,11 +15,11 @@ var (
...
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/registry/error.go:22]:
```suggestion
	return errors.New(ErrGenerateModelCode, errors.Alert, []string{fmt.Sprintf("error generating model: %s", modelName)}, []string{err.Error()}, []string{"Registrant used for the model is not supported", "Verify the model's source URL.", "Failed to create a local directory in the filesystem for this model."}, []string{"Ensure that each kind of registrant used is a supported kind.", "Ensure correct model source URL is provided and properly formatted.", "Ensure sufficient permissions to allow creation of model directory."})
```

File: mesheryctl/internal/cli/root/registry/generate.go (hunk 2 above)
Comment (leecalcote) [inline mesheryctl/internal/cli/root/registry/generate.go:225]:
```suggestion
			utils.Log.Info("Extracted ", len(comps), "components for model [ %s ]", model.ModelDisplayName)
```

File: mesheryctl/pkg/utils/component.go (hunk 5 above)
Comment (leecalcote) [inline mesheryctl/pkg/utils/component.go:156]:
```suggestion
			Log.Info(fmt.Sprintf("Reading Registrant [ %s ] Model [ %s ] Component [%s ]\n", data.Component, data.Model, data.Registrant))
```

File: mesheryctl/pkg/utils/error.go [hunk 13]
```diff
@@ -485,3 +487,19 @@ func ErrRequestResponse(err error) error {
 		[]string{"Error occurred while generating a response"},
 		[]string{"Check your network connection and the status of Meshery Server via `mesheryctl system status`."})
 }
+
+func ErrMarshalStructToCSV(err error) error {
+	return errors.New(ErrMarshalStructToCSVCode, errors.Alert,
+		[]string{"Failed to marshal struct to csv"},
+		[]string{err.Error()},
+		[]string{"Error occurred converting struct to [][]string", "The struct might have in-appropriate csv tags"},
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/error.go:495]: Rephrase. Don't talk about structs.

File: mesheryctl/pkg/utils/error.go [hunk 14]
```diff
@@ -485,3 +487,19 @@ func ErrRequestResponse(err error) error {
...
+		[]string{"Verify the value to be converted is compatbile with the columns in the spreadsheet", "Ensure the column names in the spreadsheet and struct are same"})
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/error.go:496]: Rephrase.

File: mesheryctl/pkg/utils/error.go (hunk 13 above)
Comment (leecalcote) [reply mesheryctl/pkg/utils/error.go:495]:
```suggestion
		[]string{"The column names in your spreadsheet do not match the names in the struct.", " For example, the spreadsheet has a column named 'First Name' but the struct expects a column named 'firstname'. Please make sure the names match exactly."},
```

File: mesheryctl/internal/cli/root/registry/error.go (hunk 12 above)
Comment (MUzairS15) [reply mesheryctl/internal/cli/root/registry/error.go:22]: Thanks.

File: mesheryctl/internal/cli/root/registry/generate.go (hunk 2 above)
Comment (MUzairS15) [reply mesheryctl/internal/cli/root/registry/generate.go:225]: 👍

File: mesheryctl/pkg/utils/error.go (hunk 13 above)
Comment (MUzairS15) [reply mesheryctl/pkg/utils/error.go]: Thanks

File: mesheryctl/pkg/utils/error.go (hunk 14 above)
Comment (MUzairS15) [reply mesheryctl/pkg/utils/error.go]: 👍

File: mesheryctl/pkg/utils/component.go (hunk 6 above)
Comment (MUzairS15) [reply mesheryctl/pkg/utils/component.go]: Updated

File: mesheryctl/pkg/utils/component.go (hunk 4 above)
Comment (MUzairS15) [reply mesheryctl/pkg/utils/component.go]: Updated.

**Review by leecalcote — APPROVED** [review]

**Pull request merged**